xquant/keras/keras_report_utils.py

```diff
@@ -0,0 +1,380 @@
+                for key, value in item.items():
+                    if key not in combined_dict:
+                        combined_dict[key] = []
+                    combined_dict[key].append(value)
+            for k, v in combined_dict.items():
+                combined_dict[k] = np.mean(v)
+            aggregated_metrics[layer_name] = combined_dict
+
+        return aggregated_metrics
+
+    def get_default_metrics(self) -> Dict[str, Callable]:
+        """
+        Retrieve the default metrics for evaluation.
+
+        Returns:
+            Dict[str, Callable]: A dictionary of default metric functions.
+        """
+
+        def compute_mse(f_pred: np.ndarray, q_pred: np.ndarray) -> float:
+            mse = tf.keras.losses.MeanSquaredError()(f_pred, q_pred)
+            return float(mse.numpy())
+
+        def compute_cs(f_pred: np.ndarray, q_pred: np.ndarray) -> float:
+            cs = tf.keras.losses.CosineSimilarity()(f_pred.flatten(), q_pred.flatten())
+            return float(cs.numpy())
+
+        def compute_sqnr(f_pred: np.ndarray, q_pred: np.ndarray) -> float:
+            signal_power = tf.reduce_mean(tf.square(f_pred))
+            noise_power = tf.reduce_mean(tf.square(f_pred - q_pred))
+            sqnr = signal_power / noise_power
+            return float(sqnr.numpy())
+
+        return {MSE_METRIC_NAME: compute_mse,
+                CS_METRIC_NAME: compute_cs,
+                SQNR_METRIC_NAME: compute_sqnr}
+
+    def create_float_folded_model(self,
+                                  float_model: keras.Model,
+                                  representative_dataset: Any,
+                                  ) -> keras.Model:
+        """
+        Create a folded version of the floating-point model.
+
+        Args:
+            float_model (keras.Model): The original floating-point model.
+            representative_dataset (Any): Representative dataset used during quantization.
+
+        Returns:
+            keras.Model: The folded floating-point model.
+        """
+        float_graph = self.convert_to_graph(model=float_model)
+        float_folded_model, _ = KerasImplementation().model_builder(float_graph,
+                                                                    mode=ModelBuilderMode.FLOAT,
+                                                                    append2output=None,
+                                                                    fw_info=DEFAULT_KERAS_INFO)
+        return float_folded_model
+
+    def convert_to_graph(self,
+                         model: Model,
+                         repr_dataset: Callable=None):
+        """
+        Convert Keras model to networkx graph representation.
+
+        Args:
+            model: Keras model to convert.
+            repr_dataset: Representative dataset (not used in Keras).
+
+        Returns:
+            Graph representing the model.
+        """
+        graph = graph_preparation_runner(in_model=model,
```

Comment:
rethink if we need the entire functionality in this function, maybe we can extract the relevant parts in MCT core to a separate method and just run them here

Reply:
I tried to take only parts of it.
There are two issues:
1. The need for the quantization config that determines how the optimized graph will look (like collapsing, etc.).
2. The back2fw assumes that the graph needs some quantization info **even for float models** (for example, it checks whether the node's activation should be quantized when it's not needed in the float case).

I want to change it in the future to use only the substitutions that are relevant to the float model, but it means that we need to fix two things in MCT:
1. Remove parts of the quantization config related to the graph optimizations (like collapsing, etc.).
2. Fix the back2fw to convert a float graph to the fw float model even when quantization information is missing.